src/plugins/mavlink_ftp/mavlink_ftp_impl.cpp

```diff
@@ -0,0 +1,971 @@
+    }
+
+    payload->size = 0;
+
+    return MavlinkFTP::Result::SUCCESS;
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_remove_directory(PayloadHeader *payload)
+{
+    std::string path = _root_dir + _data_as_string(payload);
+
+    if (fs::remove(path)) {
+        return MavlinkFTP::Result::SUCCESS;
+    } else {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_create_directory(PayloadHeader *payload)
+{
+    std::string path = _root_dir + _data_as_string(payload);
+
+    if (fs::create_directory(path)) {
+        return MavlinkFTP::Result::SUCCESS;
+    } else {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_remove_file(PayloadHeader *payload)
+{
+    std::string path = _root_dir + _data_as_string(payload);
+
+    if (fs::remove(path)) {
+        return MavlinkFTP::Result::SUCCESS;
+    } else {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_truncate_file(PayloadHeader *payload)
+{
+    std::string path = _root_dir + _data_as_string(payload);
+
+    std::error_code ec;
+    fs::resize_file(path, payload->offset, ec);
+    if (ec) {
+        return MavlinkFTP::Result::SUCCESS;
+    } else {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_rename(PayloadHeader *payload)
+{
+    size_t term_i = payload->size;
+    if (payload->size >= max_data_length) {
+        term_i = max_data_length - 1;
+    }
+    payload->data[term_i] = '\0';
+
+    std::string old_name = std::string(reinterpret_cast<char *>(&(payload->data[0])));
+    if (old_name.length() >= term_i) {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+    std::string new_name = std::string(reinterpret_cast<char *>(&(payload->data[old_name.length() + 1])));
+    std::error_code ec;
+    fs::rename(old_name, new_name, ec);
+    if (ec) {
+        return MavlinkFTP::Result::SUCCESS;
+    } else {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+}
+
+MavlinkFTP::Result MavlinkFTPImpl::_work_calc_file_CRC32(PayloadHeader *payload)
+{
+    uint32_t checksum = 0;
+    ssize_t bytes_read;
+    std::string path = _root_dir + _data_as_string(payload);
+
+    int fd = ::open(path.c_str(), O_RDONLY);
+
+    if (fd < 0) {
+        return MavlinkFTP::Result::ERR_FAIL;
+    }
+
+    char buffer[18392];
+
+    do {
+        bytes_read = ::read(fd, buffer, sizeof(buffer));
+
+        if (bytes_read < 0) {
+            int r_errno = errno;
+            close(fd);
+            errno = r_errno;
+            return MavlinkFTP::Result::ERR_FAIL;
+        }
+
+        checksum = crc32part((uint8_t *) buffer, bytes_read, checksum);
+    } while (bytes_read == sizeof(buffer));
+
+    close(fd);
+
+    payload->size = sizeof(uint32_t);
+    memcpy(payload->data, &checksum, payload->size);
+
+    return MavlinkFTP::Result::SUCCESS;
+}
+
+void MavlinkFTPImpl::send()
```

Comment:
Is this used anywhere?

Reply:
This part is unfinished - it is meant to be used in "streaming" download. I propose that we don't implement it for now, there are a lot of potential problems with it.

Reply:
Ok your judgement.